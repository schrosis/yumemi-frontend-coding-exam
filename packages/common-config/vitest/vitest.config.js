import { defineConfig } from "vitest/config";

/** @type {import('vitest/config').ViteUserConfig} */
const vitestConfig = {
  test: {
    coverage: {
      reporter: ["text", "json-summary", "json"],
      reportOnFailure: true,
    },
  },
};

export default defineConfig(vitestConfig);
