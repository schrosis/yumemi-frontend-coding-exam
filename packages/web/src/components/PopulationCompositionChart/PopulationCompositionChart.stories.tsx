import type { Meta, StoryObj } from "@storybook/react";
import { PopulationCompositionChart } from "./PopulationCompositionChart";

const meta: Meta<typeof PopulationCompositionChart> = {
  component: PopulationCompositionChart,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof PopulationCompositionChart>;

export const Default: Story = {
  args: {
    type: "totalPopulation",
    prefectures: [
      {
        prefectureId: 13,
        name: "東京都",
      },
      {
        prefectureId: 14,
        name: "神奈川県",
      },
    ],
    data: [
      {
        year: 2020,
        "13": {
          totalPopulation: 1000,
          youngPopulation: 500,
          workingAgePopulation: 300,
          olderPopulation: 200,
        },
        "14": {
          totalPopulation: 900,
          youngPopulation: 400,
          workingAgePopulation: 200,
          olderPopulation: 300,
        },
      },
      {
        year: 2021,
        "13": {
          totalPopulation: 1200,
          youngPopulation: 600,
          workingAgePopulation: 400,
          olderPopulation: 200,
        },
        "14": {
          totalPopulation: 1000,
          youngPopulation: 300,
          workingAgePopulation: 300,
          olderPopulation: 400,
        },
      },
      {
        year: 2022,
        "13": {
          totalPopulation: 1400,
          youngPopulation: 700,
          workingAgePopulation: 500,
          olderPopulation: 200,
        },
        "14": {
          totalPopulation: 1200,
          youngPopulation: 500,
          workingAgePopulation: 400,
          olderPopulation: 300,
        },
      },
    ],
  },
  render: (args) => (
    <div style={{ width: "100%", height: "250px" }}>
      <PopulationCompositionChart {...args} />
    </div>
  ),
};
