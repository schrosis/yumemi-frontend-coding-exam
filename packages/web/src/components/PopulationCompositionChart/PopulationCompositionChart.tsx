import {
  CartesianGrid,
  Label,
  Legend,
  Line,
  LineChart,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from "recharts";

type PopulationComposition = {
  totalPopulation: number;
  youngPopulation: number;
  workingAgePopulation: number;
  olderPopulation: number;
};

type Prefecture = {
  prefectureId: number;
  name: string;
};

export type PopulationCompositionChartProps = {
  type: keyof Omit<PopulationComposition, "year">;
  prefectures: Prefecture[];
  data: {
    year: number;
    [prefectureId: number]: PopulationComposition;
  }[];
};

export const PopulationCompositionChart: React.FC<
  PopulationCompositionChartProps
> = ({ type, prefectures, data }) => {
  return (
    <ResponsiveContainer width="100%" height="100%">
      <LineChart
        data={data}
        margin={{
          top: 30,
          right: 20,
          left: 20,
          bottom: 30,
        }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="year" padding={{ left: 0, right: 50 }}>
          <Label value="年度" offset={0} position="insideBottomRight" />
        </XAxis>
        <YAxis padding={{ top: 30, bottom: 0 }}>
          <Label value="人口数" offset={0} position="insideTopLeft" />
        </YAxis>
        <Legend />
        <Tooltip />
        {prefectures.map((prefecture) => (
          <Line
            key={prefecture.prefectureId}
            dataKey={`${prefecture.prefectureId}.${type}`}
            name={prefecture.name}
            stroke={`#${Math.floor(prefecture.prefectureId * (255 ** 3 / 100)).toString(16)}`}
          />
        ))}
      </LineChart>
    </ResponsiveContainer>
  );
};
