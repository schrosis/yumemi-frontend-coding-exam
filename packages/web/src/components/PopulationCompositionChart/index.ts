export {
  PopulationCompositionChart,
  type PopulationCompositionChartProps,
} from "./PopulationCompositionChart";
