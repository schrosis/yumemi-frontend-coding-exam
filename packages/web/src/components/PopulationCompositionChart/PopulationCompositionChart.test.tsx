import { render } from "@testing-library/react";
import { beforeAll, describe, expect, test } from "vitest";
import { PopulationCompositionChart } from "./PopulationCompositionChart";

describe("PopulationCompositionChart Component", () => {
  beforeAll(() => {
    global.ResizeObserver = class ResizeObserver {
      observe() {
        // do nothing
      }
      unobserve() {
        // do nothing
      }
      disconnect() {
        // do nothing
      }
    };
  });

  test("matches the snapshot", () => {
    const { asFragment } = render(
      <PopulationCompositionChart
        type="totalPopulation"
        prefectures={[{ prefectureId: 13, name: "東京都" }]}
        data={[
          {
            year: 2021,
            "13": {
              totalPopulation: 1000,
              youngPopulation: 500,
              workingAgePopulation: 300,
              olderPopulation: 200,
            },
          },
        ]}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
