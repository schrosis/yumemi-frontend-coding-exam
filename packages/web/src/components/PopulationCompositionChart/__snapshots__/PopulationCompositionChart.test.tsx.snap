// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopulationCompositionChart Component > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="recharts-responsive-container"
    style="width: 100%; height: 100%; min-width: 0;"
  />
</DocumentFragment>
`;
